File: prototyp/webcam/zones.py
"""THE single data contract between source and detector.

A source (mock / manual / webcam) delivers one ZoneActivity per frame with
the motion RATIOS per zone (0.0 .. 1.0). The GestureDetector reads only
.zones and .t and is therefore testable without a camera and portable to
Rust 1:1. frame/grid are optional and only meant for the overlay.
"""

from dataclasses import dataclass, field

ZONE_NAMES = ("left", "right", "up", "down")


@dataclass
class ZoneActivity:
    zones: dict            # {"left":float,"right":float,"up":float,"down":float}, each 0..1
    t: float               # timestamp in seconds (monotonically increasing)
    frame: object = None   # optional: BGR image OR mask (overlay only)
    grid: object = None    # optional: 2D bool/uint8 array of active cells (overlay only)


def full_zones(partial: dict) -> dict:
    """Fills a partial dict up to all four zones (missing -> 0.0)."""
    return {name: float(partial.get(name, 0.0)) for name in ZONE_NAMES}

File: webcam-rust/src/zones.rs
//! Data contract between source and detector (port of zones.py).
//!
//! A source delivers the motion RATIOS per zone (0.0..1.0) for each frame.
//! The detector reads only these values + the timestamp and is therefore
//! testable without a camera.

/// Ratio of active cells per zone, each 0.0..1.0.
#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct Zones {
    pub left: f64,
    pub right: f64,
    pub up: f64,
    pub down: f64,
}

impl Zones {
    pub fn new(left: f64, right: f64, up: f64, down: f64) -> Self {
        Self { left, right, up, down }
    }
}

/// One frame from the detector's point of view: zone ratios + timestamp
/// (seconds). Image/mask for the overlay are kept separately in the webcam
/// layer so this type stays free of OpenCV.
///
/// `dir_ok` is the source's direction gate for the tap zones
/// [left, right, up]: true = motion points in gesture direction (or the
/// source provides no direction info, then the gate stays open). The down
/// zone deliberately has no gate - holding means standing still, flow ~0.
#[derive(Debug, Clone, Copy)]
pub struct ZoneActivity {
    pub zones: Zones,
    pub t: f64,
    pub dir_ok: [bool; 3],
}

impl ZoneActivity {
    pub fn new(zones: Zones, t: f64) -> Self {
        Self { zones, t, dir_ok: [true; 3] }
    }

    /// Variant for sources with direction info (optical flow).
    pub fn with_dir_ok(mut self, dir_ok: [bool; 3]) -> Self {
        self.dir_ok = dir_ok;
        self
    }
}
